Vm rebuild

Ryex opened the pull request:
Goal here is to move the VM to using dynamic objects allowing the Game's logic to be  modeled much more closely.

  - [x] generic object trait
  - [x] codegen the traits, enums, etc for instructions and logic
  - [x] new frozen object interface, potential for less JSON when state is saved if the object template are in the database
  - [x] lots of interface traits to handle logic 
  - [x] load/build ItemIntergretedCircuit10 from ObjectTemplate (relatively easy)
  - [x] implementing the known circuit holders (both housings, tablet, suit)
  - [x] remap the WASM interface
  - [x] fix frontend to use new template system
  - [x] clean up new errors to reach parity with devleop

Ryex committed:
start building macros for VMObject trait

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
refactor: intorduction of a macro enabled Object trait

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
refactor(vm::objects): Establixh base object macros and traits

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
refactor(vm): use proper repr on generated enums if possible, generate stationpedia prefab enum

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex commented:
@ComputerDruid  @Emilgardis  if you want to stay away from this one I won't blame you. lots of macro work in these early stages.

Suffice it to say that I've built a rather nice and clean macro derivable `Object` interface trait that I'll be using to rework the rest of the VM. 

Ie. when I'm done as far as the VM is concerned a circuit housing is nothing more than an Object implementing the `CircuitHolder` trait and anything with memory is just an object that implements `Memory + (MemoryReadable &| MemoryWritable)`
same with logic and `Logicable`

Ryex committed:
refactor(vm): generic impls for generic Objects (not implimented with special logic)

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
refactor(vm) helpful codegen round 1~

Emilgardis commented on xtask/src/generate.rs:
```diff
@@ -0,0 +1,101 @@
+    })
+}
+
+fn prepend_genereated_comment(file_path: &std::path::Path) -> color_eyre::Result<()> {
```
Typo, `generated`

Emilgardis commented on xtask/src/generate.rs:
```diff
@@ -0,0 +1,101 @@
+    Ok(())
+}
+
+pub fn parse_json<'a, T: serde::Deserialize<'a>>(
```
Hehe, i recognize this

Ryex replied:
its' a very nice function, I wrote a much less friendly version once. I prefer this.

Emilgardis commented on xtask/src/stationpedia.rs:
```diff
@@ -0,0 +1,346 @@
+    pub description: String,
+    #[serde(rename = "Device")]
+    pub device: Option<Device>,
+    /// the item , if none then deprecated
```
My comment is wrong i think, its none if not an item. I even use that fact in https://github.com/Emilgardis/Stationeers-Wiki-Page-Helper/blob/64b21f49390ceede70d519d3dd20931a67a3a4b6/src/commands/wikibox.rs#L134

Emilgardis commented on xtask/src/main.rs:
```diff
@@ -57,6 +69,8 @@ enum Error {
     #[error("failed to run command `{0}`")]
     Command(String, #[source] std::io::Error),
+    #[error("can not find `Stationpedia.json` and/or `Enums.json` at `{0}`")]
+    BadStationeresPath(std::path::PathBuf),
```
Typo, `Stationeers`

Emilgardis commented on xtask/src/generate/instructions.rs:
```diff
@@ -0,0 +1,216 @@
+
+    write!(
+        writer,
+        "use serde::{{Deserialize, Serialize}};\n\
```
Technically, this should be serde_derive, see https://www.reddit.com/r/rust/comments/1602eah/comment/jxkciap/?utm_source=reddit&utm_medium=web2x&context=3

Ryex replied:
I'm not sure I follow the repo's read-me or the reddit post... it it saying that I shouldn't use the reexport because it allows the possibility of version conflicts?

Emilgardis replied:
it's just a best practice that is not followed in the ecosystem currently, the idea is that if you use `serde_derive` for the proc macros, crates only using serde without derive can run earlier since it doesn't need syn+quote

Emilgardis commented on xtask/src/stationpedia.rs:
The structs here technically dont need to be `deny_unknown_fields` especially since serde_ignored is used

Emilgardis commented on ic10emu/src/vm/object/mod.rs:
```diff
@@ -0,0 +1,54 @@
+#[derive(Debug, Default, Clone, Serialize, Deserialize)]
+pub struct Slot {
+    pub typ: SlotType,
+    pub enabeled_logic: Vec<SlotLogicType>,
```
Typo, `enabled`

Emilgardis commented on xtask/Cargo.toml:
```diff
@@ -5,4 +5,22 @@ edition.workspace = true
+tracing = "0.1.40"
+tracing-subscriber = { version = "0.3.18", features = ["env-filter"] }
+
+onig = { git = "https://github.com/rust-onig/rust-onig", revision = "fa90c0e97e90a056af89f183b23cd417b59ee6a2" }
```
Why onig and not `regex` crate ?

Ryex replied:
because I needed lookahead to correctly handle nested "color" tags in the regex. if you have come up with an alternative in the Wiki helper I'm all ears, `onig` is kinda broken atm, hence the git version and crates-io patch.

Emilgardis commented on xtask/src/generate/database.rs:
```diff
@@ -0,0 +1,674 @@
+) -> color_eyre::Result<()> {
+    let templates = generate_templates(stationpedia)?;
+
+    println!("Writing prefab database ...");
```
This should be eprintln or use tracing

Emilgardis commented on xtask/src/generate/instructions.rs:
```diff
@@ -0,0 +1,216 @@
+    writer: &mut T,
+    instructions: &BTreeMap<String, stationpedia::Command>,
+) -> color_eyre::Result<()> {
+    println!("Writing instruction Listings ...");
```
Ditto, eprintln or tracing

Emilgardis commented on xtask/Cargo.toml:
```diff
@@ -5,4 +5,22 @@ edition.workspace = true
+textwrap = { version = "0.16.1", default-features = false }
 thiserror = "1.0.58"
+tracing = "0.1.40"
+tracing-subscriber = { version = "0.3.18", features = ["env-filter"] }
```
Unused dep

Emilgardis reviewed (commented):
Looks great!

Ryex committed:
refactor(vm) cleanup errors

Ryex committed:
refactor(vm) cleanup serde_derive usage

Ryex committed:
refactor(vm) use generated enums wherever possible

Emilgardis commented on xtask/src/generate/instructions.rs:
```diff
@@ -0,0 +1,215 @@
+        use crate::vm::object::traits::{{Logicable, MemoryWritable, SourceCode}};\n\
+        use crate::errors::ICError; \n\
+        pub trait IntegratedCircuit: Logicable + MemoryWritable + SourceCode {{\n    \
+            fn get_instruciton_pointer(&self) -> usize;\n    \
```
typo, `instruction`

Ryex committed:
refactor(vm) checkup - ensure tests pass

Ryex committed:
refactor(vm) start impl of IC trait function

Ryex committed:
refactor(vm) start reimpl of VM struct using object trait

Ryex committed:
refactor(vm): start freezeing objects impl

Ryex committed:
refactor(vm): use From<T> in freeze object

Ryex committed:
refactor(vm): VMObjects should carry around a VM reference

Ryex committed:
chore(vm): clean up spelling and RC usage

Ryex committed:
refactor(vm): re impl freezing VM using `ObjectTemplate`s

Ryex committed:
refactor(vm): update up codegen, remove duplicated enums

Ryex committed:
refactor(vm): more ic10 instruction impls

Ryex commented:
phew... finally getting close to parity after the rewrite. so much code... 
once I finish off porting the instructions to traits it will just be a case of  
  - [x] load/build ItemIntergretedCircuit10 from ObjectTemplate (relatively easy)
  - [x] implementing the known circuit holders (both housings, tablet, suit)
  - [ ] remap the WASM interface
  - [ ] fix frontend to use new template system
  
  at this point the VM and object system is stable, so if anyone wants to double check it for my sanity that would be great.

Ryex committed:
rafactor(vm): end instruction impl

Ryex committed:
refactor(vm): cleanup borrow usage, impl slot logic

Ryex committed:
refactor(vm): impl circuit holder, impl ObjectRef/Mut

Ryex committed:
chore(clippy): sweep sweep sweep

Ryex committed:
refactor(vm): pull generated data into new crate

pregenerate prefab templates -> locked behind crate feature

Ryex committed:
refactor(vm): update prefab database with new data

Ryex committed:
refactor(vm): drop null fields from json

Ryex committed:
refactor(vm): rework object freazing

 - let vm carry object dtabase
 - seperate frozen object info and templates

Ryex committed:
refactor(vm): generic circuit holder + suit impl

Ryex committed:
refactor(vm): freeze concreet ic, add "humans", tsify feature

Ryex committed:
refactor(vm): make wasmbindings build again, use tsify for types

Ryex commented:
*manic laughing commences*
@Emilgardis 
@ComputerDruid 
It's Done.
the Rust rebuild is complete.

I now have to remap the wasm bindings and fix the frontend

Emilgardis commented:
😵

Emilgardis commented on ic10emu/src/vm/instructions/traits.rs:
```diff
@@ -0,0 +1,4615 @@
+        &mut self,
+        r: &crate::vm::instructions::operands::Operand,
+        a: &crate::vm::instructions::operands::Operand,
+    ) -> Result<(), crate::errors::ICError> {
```
should probably `use crate::vm::instructions::operands::{Operand, InstOperand};` in this module to skip all the qualified paths

Ryex replied:
this function uses Operand because it and it's call is generated. it performs the transformation to `InstOperand` and calls `execute_inner` which is hand written.

Emilgardis replied:
The reason to include it would be to save on bytes and cognitive load, there's no reason to always qualify here

Ryex replied:
.. I think my brain may have skipped when I first read that. for some reason I was thinking you were suggesting that converting the function to use InstOperand. but yeah, your correct I think.

Emilgardis commented on ic10emu/src/vm/instructions/enums.rs:
```diff
@@ -0,0 +1,1567 @@
+use serde_derive::{Deserialize, Serialize};
```
missing a big disclaimer that this is code-gen, I still feel like it should be put in a path like `src/vm/instructions/codegen/` so that it's clear by just lookit at the file that it's code-gen, instead of only having to scroll up and seeing a disclaimer.

Same for traits.rs

Ryex replied:
apparently `quote` does not preserve comments. fixed

Emilgardis commented on ic10emu/src/vm/object/stationpedia/structs/circuit_holder.rs:
```diff
@@ -0,0 +1,417 @@
+}
+
+#[allow(dead_code)]
+impl StructureCircuitHousing {
```
not needed

Emilgardis commented on ic10emu/src/vm/object/errors.rs:
```diff
@@ -0,0 +1,38 @@
+use wasm_bindgen::prelude::*;
+
+#[derive(Error, Debug, Clone, Serialize, Deserialize)]
+#[cfg_attr(feature = "tsify", derive(Tsify))]
+#[cfg_attr(feature = "tsify", tsify(into_wasm_abi, from_wasm_abi))]
```
this is mostly just styling preference, but it's possible to do this like. I have no preference


```suggestion
#[cfg_attr(feature = "tsify", derive(Tsify), tsify(into_wasm_abi, from_wasm_abi))]
```

Emilgardis commented on ic10emu/src/vm/object/stationpedia.rs:
```diff
@@ -0,0 +1,152 @@
+                        .clone()
+                        .unwrap_or_else(|| prefab.get_str("name").unwrap().to_string())),
+                ),
+                prefab: Name::from_prefab_name(&prefab.to_string()),
```
```suggestion
                prefab: Name::from_prefab_name(prefab.as_ref()),
```

Emilgardis commented on ic10emu/src/vm/object/stationpedia.rs:
```diff
@@ -0,0 +1,152 @@
+                    .map(TryInto::try_into)
+                    .transpose()
+                    .map_err(|vec: Vec<f64>| TemplateError::MemorySize(vec.len(), 512))?
+                    .unwrap_or_else(|| [0.0f64; 512]),
```
```suggestion
                    .unwrap_or([0.0f64; 512]),
```

Emilgardis commented on ic10emu/src/vm/object/stationpedia.rs:
```diff
@@ -0,0 +1,152 @@
+                    .transpose()
+                    .map_err(|vec: Vec<f64>| TemplateError::MemorySize(vec.len(), 18))?
+                    .unwrap_or_else(|| [0.0f64; 18]),
+                ip: obj
```
```suggestion
                    .unwrap_or([0.0f64; 18]),
```

Emilgardis commented on ic10emu/src/vm/object/templates.rs:
```diff
@@ -0,0 +1,1409 @@
+    }
+}
+
+impl ObjectInfo {
```
should probably document these functions

Ryex replied:
👍

Emilgardis commented on ic10emu/src/vm/object/traits.rs:
```diff
@@ -0,0 +1,364 @@
+        fn get_radiation_factor(&self) -> f32;
+    }
+
+    pub trait IntegratedCircuit: Logicable + MemoryWritable + SourceCode + Item {
```
some documentation on these functions would be nice

Ryex replied:
Is this what you were looking for? or should I be more in depth?

Emilgardis replied:
looks good to me!

Emilgardis commented on ic10emu/src/grammar.rs:
```diff
@@ -1090,21 +668,35 @@ impl Number {
+
     use super::*;
 
+    static INIT: std::sync::Once = std::sync::Once::new();
```
shouldn't this be a thread local?

Ryex replied:
can tests get multi-threaded? and would that matter? I'm using it to set color_eyre's panic hook.

Emilgardis replied:
I was thinking about tracing, never mind!

Emilgardis commented on ic10emu/src/interpreter.rs:
```diff
@@ -2626,80 +214,87 @@ pub fn i64_to_f64(i: i64) -> f64 {
-    use crate::vm::VMError;
-
-    use super::*;
 
```
don't forget these tests!

Ryex replied:
opp! your right. I commented them out during the rebuild because the interface changed a lot.
Fixed and passing.

Ryex committed:
refactor(vm): cleanup, trait docs, reenable interpreter tests

Ryex committed:
refactor(wasm): remap wasm interface (rust side)

Ryex committed:
chore(codegen): generate json database in www/data

Ryex committed:
refactor(frontend): use new frozen object to fill values for baseObject mixin, proxy vm to webworker to avoid hogging main thread

Ryex committed:
refactor(vm, frontend): memory instructions, baseObject

- parse and map meory instructions
- use FrozenObjectFull to propogate  data out of VM to componates (far
  less wasm calls)

Ryex committed:
refactor(frontend): fix base template component

Ryex committed:
refactor(frontend): fix slot occupant and device add components

Ryex committed:
refactor(vm, frontend): internally tag template enum, finish remap frontend

Ryex committed:
refactor(vm, frontend): turns out that the serde_wasm_bindgen facilities of TSify are borken...

Ryex committed:
refactor(frontend): Maps are dea dlong live Objects, also delay main body render untill VM is fully loaded

Ryex committed:
refactor(ui): start using signals

smaller DOM updates  & easier data update paths.
fix sl-select problems with force update on signal sub

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
refactor(frontend) finish signal conversion, fix passing data to VM

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
refactor(frontend): fix signal graph

Ryex committed:
refactor(frontend) fix template searching, update database images

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
refactor(frontend): fix loading code to editor

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
refactor(vm): add proper reagent interface, trace VM, fix slot serialization

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
refactor(vm): fix Reagent code gen

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
refactor(frontend): use signals in Session, fix various things

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
refactor(vm+frontend): better logging, fix slot serialization

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
refactor(frontend): bump rebuild to v1.0

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>
